src/elf/gnu_hash.rs

```diff
@@ -1,143 +1,206 @@
 //! A Gnu Hash table as 4 sections:
 //!
-//!   1. Header
-//!   2. Bloom Filter
-//!   3. Hash Buckets
-//!   4. Hash Values
+//!  1.Header
+//!  2.Bloom Filter
+//!  3.Hash Buckets
+//!  4.Chains
 //!
 //! The header has is an array of four (4) u32s:
 //!
-//!   1. nbuckets
-//!   2. symndx
-//!   3. maskwords
-//!   4. shift2
+//!  1.nbuckets
+//!  2.symndx
+//!  3.maskwords
+//!  4.shift2
 //!
-//! See: https://blogs.oracle.com/solaris/gnu-hash-elf-sections-v2
+//! See more:
+//!  * http://www.linker-aliens.org/blogs/ali/entry/gnu_hash_elf_sections
+//!    or https://blogs.oracle.com/solaris/gnu-hash-elf-sections-v2
+//!  * https://flapenguin.me/2017/05/10/elf-lookup-dt-gnu-hash/
 
-/// GNU hash function: takes a string and returns the u32 hash of that string
+/// GNU hash function: accepts a symbol name and returns a value that may be
+/// used to compute a bucket index.
+///
+/// Consequently, if the hashing function returns the value `x` for some name,
+/// `buckets[x % nbuckets]` gives an index, `y`, into both the symbol table
+/// and the chain table.
 pub fn hash(symbol: &str) -> u32 {
     const HASH_SEED: u32 = 5381;
-    let mut hash = HASH_SEED;
-    for b in symbol.as_bytes() {
-        hash = hash
-            .wrapping_mul(33)
-            .wrapping_add(u32::from(*b));
-    }
-    hash
+    symbol.bytes().fold(HASH_SEED, |hash, b| {
+        hash.wrapping_mul(33).wrapping_add(u32::from(b))
+    })
 }
 
 #[cfg(test)]
 mod tests {
     use super::hash;
     #[test]
     fn test_hash() {
-        assert_eq!(hash("")             , 0x00001505);
-        assert_eq!(hash("printf")       , 0x156b2bb8);
-        assert_eq!(hash("exit")         , 0x7c967e3f);
-        assert_eq!(hash("syscall")      , 0xbac212a0);
+        assert_eq!(hash(""), 0x00001505);
+        assert_eq!(hash("printf"), 0x156b2bb8);
+        assert_eq!(hash("exit"), 0x7c967e3f);
+        assert_eq!(hash("syscall"), 0xbac212a0);
         assert_eq!(hash("flapenguin.me"), 0x8ae9f18e);
     }
 }
 
 macro_rules! elf_gnu_hash_impl {
     ($size:ty) => {
-
-        use core::slice;
-        use core::mem;
-        use crate::strtab::Strtab;
         use super::sym;
+        use crate::strtab::Strtab;
+        use core::mem;
+        use core::slice;
+
+        const BYTE_SIZE: u32 = 8;
+        /// Size of a bits mask in bloom filter
+        const ELFCLASS_BITS: u32 = mem::size_of::<$size>() as u32 * BYTE_SIZE;
 
-        pub struct GnuHash<'process> {
+        pub struct GnuHash<'a> {
```

Comment:
thank you for documenting all these fields! Do note this is another good example though that makes this hard to read, almost every field changed in this struct...

Reply:
Sorry about that :)